docs: Add VM latency checkup design document

orelmisan opened the pull request:
This adds a design document for the VM latency checkup.

Signed-off-by: Orel Misan <omisan@redhat.com>

EdDev commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,334 @@
+
+| Env Variable Name                | Description                                                                                      | Is Mandatory | Remarks                                                              |
+|----------------------------------|--------------------------------------------------------------------------------------------------|--------------|----------------------------------------------------------------------|
+| NAD_NAMESPACE                    | Namespace where the NetworkAttachmentDefinition object resides.                                  | True         |                                                                      |
```
As @phoracek pointed out, we should avoid the use of the name `NAD`. Lets use the full name.

orelmisan replied:
Done.

EdDev commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,334 @@
+| SOURCE_NODE                      | Check latency from this node to TARGET_NODE                                                      | False        | When specified the VM will start with a corresponding affinity rule. |
+| TARGET_NODE                      | Check latency from SOURCE_NODE to this node                                                      | False        | When specified the VM will start with a corresponding affinity rule. |
+| SAMPLE_DURATION_SECONDS          | Latency check duration [sec]                                                                     | False        | Defaults to 5 [sec]                                                  |
+| MAX_DESIRED_LATENCY_MILLISECONDS | Maximum desired latency between VMs. If the latency is higher than this - the checkup fails [ms] | False        | Defaulting to 0. 0 translates to a no-max-latency limit.             |
```
Please try to see if the columns can be shrink a bit (splitting tests into more lines).

EdDev reviewed (changes requested):
Thank you, looks great.
Just a few comments inline.

orelmisan commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,334 @@
+
+### Input
+
+**Note**: The K8s checkup framework provides additional environment variables (for example the output ConfigMap).
```
Done.

orelmisan commented:
Change: Replaced "NAD" with NetworkAttachmentDefinition".

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/e2aaad1c812d43057d41789fd1121db2be77d334..8afde099232eb886c12220576f5d76a6b1959dd0): Changed failureReason to be an empty string on success.

ormergi commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,333 @@
+1. Check if Kubevirt is installed and ready.
+2. Check if Multus exists.
+3. Check if the input NetworkAttachmentDefinition exists.
+4. There are at least two worker nodes.
```
I think it should also check if source and/or target nodes the user required exists.
It will probably require nodes viewer permissions.

orelmisan replied:
@ormergi is it possible to limit the data that the checkup will receive to just the Nodes names?

ormergi replied:
Looking at this again I find it not clear, what I meant is that it should check if source and/or target nodes that were specified exists (not as an additional setp to other checks).

@orelmisan I dont think so, AFAIK RBAC API doesn't provide such granularity (up to object fields).
Note that if the nodes are not exists the checkup will eventually fail when the VMs are created,
having such check will make the checkup fail faster.

EdDev replied:
We have learned that security/permissions is prioritized over fast failure.
@orelmisan , perhaps you can adjust it to be more generic... e.g. `The required nodes are available to run the checkup.`

This should cover implicit and explicit nodes specifications, even futuristic selectors can fall into this spec.

orelmisan replied:
@EdDev I've updated section 4 according to your suggestion.

orelmisan commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/37d0d119bdccb7e05405cba165a68a027ee89705..1f4ff873ca07a92c63f1631eb676ae590ca7572d): Added `status.result.sourceNode` and `status.result.targetNode` in order to address issue #113.
Changed `framework` to `Kiagnose`.
Fixed similar issues that came up during the review of PR #11.

RamLavi reviewed (approved):
/lgtm

EdDev reviewed (approved):
Thank you!

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/7384c337012c92950f1a6e0722d1c52d3dee0cb3..97ca33f9a303eb7a647f0a869286ef143128bd94): Rebase
Changed block diagram to match the main design document.

RamLavi commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,329 @@
+
+The checkup shall perform the following:
+
+1. Check if the given NetworkAttachmentDefinition makes use of the SR-IOV CNI plugin.Create two VMI’s, each with node affinity according to the
```
didn't we change it so that we can wither use SRIOV or linux-bridge cni? If so, then we need to be accurate here.

orelmisan replied:
@RamLavi thanks for catching that.
I've changed the wording a little bit to make it more agnostic to CNI type.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/97ca33f9a303eb7a647f0a869286ef143128bd94..b48a9b4969cd705a85336bf287647528f782fbc3): Answered a [comment](https://github.com/kiagnose/kiagnose/pull/12#discussion_r897860522).

RamLavi commented:
/lgtm

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/b48a9b4969cd705a85336bf287647528f782fbc3..d6c8cd6de129c2961916316476812f79bee4fca0): Changed wording in checkup's setup as a result from offline comment by @EdDev.

ormergi commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,326 @@
+
+1. Check the connectivity and latency of a network in my cluster, by:
+
+   1.1. Having an active cluster, with at least two worker nodes.
```
```suggestion
   1.1. Having an active cluster, with at least one worker node.
```
Having more then one worker is not a must, one may want to check connectivity between two VMs on the same node.

EdDev replied:
I do not think so.
This is a requirement placed here explicitly and if the checkup is not making sure this is the case (e.g. use anti-affinity) then we have a problem/bug. If a user will specify the same node explicitly, I can accept it, but it should be an exception and not the common/recommended case.

ormergi commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,326 @@
+
+   1.1. Having an active cluster, with at least two worker nodes.
+
+   1.2. Defining, configuring and creating a NetworkAttachmentDefinition, which I am interested in checking.
```
```suggestion
   1.2. Create a NetworkAttachmentDefinition object of the network I am interested in checking.
```

ormergi commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,326 @@
+
+As a Kubernetes Cluster Administrator I would like to:
+
+1. Check the connectivity and latency of a network in my cluster, by:
```
1.1 and 1.2 seem to be requirements and not user stories, is it on purpose?
I think 1.2 and 1.3  can be combined, 1.3 seems redundant.

orelmisan replied:
This document is a 1 to 1 conversion of the original google doc file to markdown.
I suggest to look there for the reasoning.

ormergi replied:
The original not only this doc is not opened for anyone so referring to it for reasoning is not effecting.
Other than that, the reasoning for what I ask is not mentioned there.

ormergi commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,326 @@
+
+   1.2. Defining, configuring and creating a NetworkAttachmentDefinition, which I am interested in checking.
+
+   1.3. Running the Kubevirt Network Latency checkup, while providing as input the NetworkAttachmentDefinition.
```
```suggestion
   1.3. Running the Kubevirt Network Latency checkup, while providing the `NetworkAttachmentDefinition` as input.
```

orelmisan replied:
Done.

ormergi commented on docs/vm_latency_checkup_design.md:
```diff
@@ -0,0 +1,326 @@
+
+## Deliverables
+
+According to Kiagnose requirements.
```
This seems not effective, it should mention at least a container-image and documentation.
Alternatively this section can be added after Kiagnose docs are merged and add a ref.

orelmisan replied:
Added a link to the requirements.

ormergi reviewed (commented):
There are similar topics from #11 I think should be also addressed here https://github.com/kiagnose/kiagnose/pull/11#pullrequestreview-1011522535, other than that please see my inline comments.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/d6c8cd6de129c2961916316476812f79bee4fca0..352aa39bbba06aef2b4fdb1ac53217f67e396808): Updated yaml manifests.
Fixed note rendering.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/577965509fd7b6e1f0f4e382213de99dfcd71ce4..d85a0a0b3902512abf3406f1074bf0d291353073): Answered a [comment](https://github.com/kiagnose/kiagnose/pull/12#discussion_r901106294).

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/d85a0a0b3902512abf3406f1074bf0d291353073..4207f5bf0009c1f96774a437f021e32c8c319895): Answered a [comment](https://github.com/kiagnose/kiagnose/pull/12#discussion_r901120499).

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/4207f5bf0009c1f96774a437f021e32c8c319895..dba4f37cbb9e47ece32850a1a19af44d472b7986): Answered @ormergi offline comments.

ormergi reviewed (approved):
Thanks for the changes, LGTM.
